mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/support/handler/AckHandler.java

```diff
@@ -14,12 +14,22 @@
 package io.streamnative.pulsar.handlers.mqtt.support.handler;
 
 import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.mqtt.MqttQoS;
 import io.streamnative.pulsar.handlers.mqtt.Connection;
+import io.streamnative.pulsar.handlers.mqtt.messages.factory.MqttSubAckMessageHelper;
+import java.util.List;
 
 /**
  * Ack handler.
  */
 public interface AckHandler {
 
     ChannelFuture sendConnAck(Connection connection);
+
+    ChannelFuture sendSubAck(Connection connection, int packetId, List<MqttQoS> grantedQoses);
```

Comment:
We'd better define only one method : sendSubAck like ChannelFuture sendSubAck(Connection connection, SubscribeAck subAck);  And SubscribeAck contains all the fields(packetId, grantedQoses, errorReason, reasonStr) that process all the cases.

Reply:
fixed